From 4017654ebac792e6c72646953de637d028e2250a Mon Sep 17 00:00:00 2001
From: Thanneeru Srinivasulu <tsrinivasulu@cavium.com>
Date: Wed, 31 May 2017 15:04:57 +0530
Subject: [PATCH 156/375] net: thunderx: Don't disable auto-negotiation for KR
 interfaces

For 10G-KR and 40G_KR interfaces don't disable auto-negotiation.
Firmware will takecare of enabling or disabling of the same
based on PHY present on-board.

Signed-off-by: Thanneeru Srinivasulu <tsrinivasulu@cavium.com>
Signed-off-by: Sunil Goutham <sgoutham@cavium.com>
---
 drivers/net/ethernet/cavium/thunder/thunder_bgx.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/cavium/thunder/thunder_bgx.c b/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
index c8a16b0347d7..179c2b6b2d2f 100644
--- a/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
+++ b/drivers/net/ethernet/cavium/thunder/thunder_bgx.c
@@ -613,11 +613,15 @@ static int bgx_lmac_xaui_init(struct bgx *bgx, struct lmac *lmac)
 	cfg &= ~SPU_FEC_CTL_FEC_EN;
 	bgx_reg_write(bgx, lmacid, BGX_SPUX_FEC_CONTROL, cfg);
 
-	/* Disable autoneg */
+	/* Enable autoneg for KR interfaces */
 	cfg = bgx_reg_read(bgx, lmacid, BGX_SPUX_AN_CONTROL);
-	cfg = cfg & ~(SPU_AN_CTL_AN_EN | SPU_AN_CTL_XNP_EN);
-	bgx_reg_write(bgx, lmacid, BGX_SPUX_AN_CONTROL, cfg);
+	cfg = cfg & ~SPU_AN_CTL_XNP_EN;
+	if (lmac->use_training)
+		cfg |= SPU_AN_CTL_AN_EN;
+	else
+		cfg &= ~SPU_AN_CTL_AN_EN;
 
+	bgx_reg_write(bgx, lmacid, BGX_SPUX_AN_CONTROL, cfg);
 	cfg = bgx_reg_read(bgx, lmacid, BGX_SPUX_AN_ADV);
 	if (lmac->lmac_type == BGX_MODE_10G_KR)
 		cfg |= (1 << 23);
@@ -629,7 +633,12 @@ static int bgx_lmac_xaui_init(struct bgx *bgx, struct lmac *lmac)
 	bgx_reg_write(bgx, lmacid, BGX_SPUX_AN_ADV, cfg);
 
 	cfg = bgx_reg_read(bgx, 0, BGX_SPU_DBG_CONTROL);
-	cfg &= ~SPU_DBG_CTL_AN_ARB_LINK_CHK_EN;
+
+	if (lmac->use_training)
+		cfg |= SPU_DBG_CTL_AN_ARB_LINK_CHK_EN;
+	else
+		cfg &= ~SPU_DBG_CTL_AN_ARB_LINK_CHK_EN;
+
 	bgx_reg_write(bgx, 0, BGX_SPU_DBG_CONTROL, cfg);
 
 	/* Enable lmac */
-- 
2.14.1

